elastic/kibana
# [Synthetics Service] Add warnings for when service nodes bandwidth is exceeded

> ⚠️ **WARNING**: Please notice that part of this functionality (latency alerts) should _not_ be shown anymore as per the following PR: https://github.com/elastic/kibana/pull/129597. Make sure to test the requirements in that PR instead of this one.

## Summary

This PR addresses the Synthetics Service issue about network bandwidth limits. I can't link it here as it's on a non-public repo.

It does so by adding warnings for when the user exceeds throttling limits, or does not enable throttling when assigning monitors to a location.

![Screenshot 2022-03-17 at 11 24 38](https://user-images.githubusercontent.com/6868147/158799294-3e69774c-e468-45c4-8e06-080c0fd456d6.png)

![Screenshot 2022-03-17 at 11 24 27](https://user-images.githubusercontent.com/6868147/158799305-572ce051-18c0-4041-9bd2-7681e040cebd.png)

## How to test this PR Locally

1. Create a new `manifest.json` with the following content and place it under a folder named, for example, `manifest_folder`.
    ```json
    {
      "throttling": {
        "downloadBandwidthLimit": 20,
        "uploadBandwidthLimit": 10,
        "latencyLimit": 5000
      },
      "locations": {
        "Local Synth Node [MANIFEST]": {
          "url": "https://localhost:10001",
          "geo": {
            "name": "Example - Local",
            "location": {"lat": 41.25, "lon": -95.86}
          },
          "status": "beta"
        }
      }
    }
    ```
2. Run an HTTP server to serve the contents of the `manifest_folder`
    ```
    $ npx http-server manifest_folder --port 8082
    ```
3. In Kibana's config, fill the following options:
    ```yaml
    # Make sure your pods will be able to send data to ES which should be bound to your host's 9200 port
    xpack.uptime.service.hosts: [http://host.docker.internal:9200]
    
    # Add the URL for the manifest with the `throttling` configs you're serving
    xpack.uptime.service.manifestUrl: http://localhost:8082/manifest.json

    # Make sure to disable the devUrl so you're only fetching locations from the `manifest` as we'll do in production
    # xpack.uptime.service.devUrl: https://localhost:10001

    # Enable the Monitor Management UI
    xpack.uptime.ui.monitorManagement.enabled: true
    ```
4. Run the service locally
    ```
    $ DEV=true LOGLEVEL=trace CACERTFILE=./k8s/cert-test/build/bundle.pem SERVERCERT=./k8s/cert-test/build/local-server.pem SERVERKEYFILE=./k8s/cert-test/build/local-server.key PORT=10001 ./synthetics-service
    ```
5. Run Kibana and try to add a monitor. You should see the warnings above when exceeding the limits defined in your `manifest.json` file.
6. Disable throttling and you should see a warning about the automatic caps.
7. Create a monitor with a particular throttling value set and save it. Use the following code for that monitor:
    ```js
    step("load homepage", async () => {
        await page.goto('https://www.fast.com');
    });
    
    step("wait for test", async () => {
      await page.waitForSelector('#show-more-details-link');
      await page.waitForTimeout(10000);
      await page.click("#show-more-details-link");
    });
    
    step("wait for up/lat", async () => {
      await page.waitForSelector(".extra-measurement-value-container.succeeded");
      await page.waitForTimeout(10000);
    });
    ```
8. Wait for that monitor to run and ensure your throttling values have been applied.
9. Create a monitor with throttling disabled using the same script as above and confirm that your script has not been throtted.
    ⚠️ It will be throttled on the service because of platform limits.
10. Create a monitor with throttling values which exceed the limits for the node. You should still be able to save such monitors.

### Checklist

Delete any items that are not applicable to this PR.

- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials -> **Do we need this now or should we wait until we're out of beta?**


### For maintainers


## Release note

Show warning when users exceed a Synthetics Node throttling limits.
